### PR TITLE
feat: add `JSXSpreadChild` and tool to build keys out of AST definitions

### PR DESCRIPTION
**Pull request by brettz9**
Refiling with modifications from original PR #35.

This PR adds:
1. chore: tool to build keys out of AST definitions
1. In the process of making these changes (and using the approach advised by @mdjermanovic to look at interface `type` rather than name), the following are currently still changed:
    1. ~~Breaking Change: Removes `ExperimentalRestProperty`, `ExperimentalSpreadProperty`~~
    2. feat: Adds `JSXSpreadChild`
1. refactor: Alphabetizes the Node types for easier comparisons (though the ordering of their keys remain unchanged). (Could be breaking if Node type iteration order were significant though I wouldn't expect this to be the same issue as with keys since the Node types seem to be more of a map.)

Here are the items on which I think I in particular still need feedback or action:
- [x] Whether to force-add the `Experimental*` properties or just let them be dropped as per the above (and confirm that `JSXSpreadChild` should indeed be added).

- [x] I currently allow two approaches in the builder toward restricting output: suppressing by a given key name or by making an interface non-traverseable. For example, `loc`, `comments`, `innerComments`, and `operator` are currently added to the properties suppressed by `getKeys` (`parent`, `leadingComments`,  and `trailingComments`) in order to ensure they continue not to appear in the listing. (This also raises another question. See next item.) I could alternatively suppress the `SourceLocation`, `Comment`, and `*Operator` types from being treated as traverseable (and thus properties of these types would not be added as keys). But even if it doesn't matter the approach, I still need to know if my suppressing those properties was suitable.

Besides the impact on keys, the Nodes that are added are affected. I'm currently not including `SourceLocation` (it is not listed as an exportable interface, doesn't have a type, and was not included previously), nor am I including `Line` and `Block` type comment Nodes despite their having exportable AST available, as this had not previously been included (and they don't seem to be accepted by ESTree), though I can do so if desired. See also discussion below on comments.

- [x] Whether `getKeys` needs to suppress the above-mentioned keys also, and also suppress `type` since the Node argument supplied by the user presumably may have a `type` on it.

While it would seem to me from my limited perspective that it would be better to err on the side of including the comment info, as I imagine people might have a desire to traverse or query this separately, if that were to be the case, it raises the ESTree question for me of why non-block nodes could not have comment properties; in TypeScript, for example, such inline comments are often used for [casting](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#casts). Similarly, I would think it would be ideal to have a standard for JSDoc-block attachment even if as a mere optional extension to the standard, e.g., an optional `jsdoc` property associated with `JSDocBlock`, `JSDocTag`, the various types, etc. (as we do with https://github.com/es-joy/jsdoccomment/blob/main/src/commentParserToESTree.js#L237-L242 and https://github.com/jsdoc-type-pratt-parser/jsdoc-type-pratt-parser/blob/34bd2a6e72af6b16b47e4e1a4391142035917e89/src/visitorKeys.ts#L7-L35 or as it appears TypeScript may be doing: https://github.com/typescript-eslint/typescript-eslint/blob/550b61ee1096113b234bf035dd267ba385809961/packages/typescript-estree/src/ts-estree/ts-nodes.ts#L184-L208 ). I could file this at ESTree, but mentioning here as the PR raises the question.

- [x] I'll wait on adding full build tool coverage until knowing the final expectations.

### CODE REVIEW

**Comment by brettz9** [comment]
I'm not clear on why the CLA is failing as GitHub shows my user next to the commit.

**Comment by nzakas** [comment]
Double check that the email associated with the commit is one associated with your GitHub user account.

File: tools/get-keys-from-ts.js
Comment (mdjermanovic) [inline tools/get-keys-from-ts.js:190]: Would it make sense to get the order of keys from `lib/visitor-keys.js` for known nodes, and just add new nodes (and new keys for known nodes) with any order of keys? That would avoid updating this logic whenever there's something new that isn't covered by it.

Comment (brettz9) [reply tools/get-keys-from-ts.js:190]: I followed this approach but making it just a little more transparently deterministic, I added a commit to alphabetize the remainder at the end of any existing keys.

File: lib/visitor-keys.js
Comment (brettz9) [reply lib/visitor-keys.js:187]: Ok, sure. Whatever is decided for any additions or removals, I can just make exceptions as needed.

Comment (nzakas) [reply lib/visitor-keys.js:187]: I think it’s safe to add here.

File: lib/visitor-keys.js
Comment (nzakas) [reply lib/visitor-keys.js:102]: I think we should leave these. It doesn’t hurt to leave in but we could potentially break existing users without an ESLint major upgrade.

**Comment by brettz9** [comment]
Ok, so to break down the next question of my OP perhaps a little more digestibly:

Based on the previous pattern of our keys, I understand that we do not want `loc`, `comments`, `innerComments`, and `operator` to show up, even though they do have, like other properties we do want, their own interface (i.e., `Comment`  and `SourceLocation`) or type (the `*Operator` string union types like `UnaryOperator`) and as such would normally be expected to be traversable by the algorithm I have set up.

I didn't settle definitively on one or the other. I can go solely with either of a few approaches of determining how these properties should be ignored and would like to know which one or ones you all think is best for me to use:

1. Explicitly suppressing the undesired `loc`, `comments`, `innerComments`, and `operator` properties above and beyond those suppressed by `getKeys`.
2. Adding `loc`, `comments`, `innerComments`, and `operator` to the current explicit exclusions made by `getKeys` (i.e., excluding these new ones along with the existing `parent`, `leadingComments`,  and `trailingComments` that comprise the `KEY_BLACKLIST`).
3. Suppress the `SourceLocation` and `Comment` interfaces, and `*Operator` types from being treated as traverseable (and thus properties of these would not be added as keys)

Currently, the results would be the same. but it seems, all things being equal, that it may be more conceptually appealing to use one consistent approach.

No. 1 and 2 seems they would be riskier if these names might legitimately need traversal in some new interface. Although no. 3 seems the most satisfying to me personally as it confines the exclusions not by property name but by interfaces considered not to be traversed, no. 2 seems most consistent with our current approach, given that `getKeys` is excluding specific properties regardless of the node type that contains them. 

But even if choosing no. 3, I wonder whether no. 2 should be implemented so that our `getKeys` also excludes the undesired properties when encountered. I could do so by autogenerating our `KEY_BLACKLIST` as well as our KEYS (though `parent` would need to be manually added since the AST doesn't specify `parent` so as to allow it to be programmatically detected somehow; the other two currently in the blacklist, `leadingComments` and `trailingComments`, could be excluded based on their pointing to an array of the undesired `Comment`).

File: tools/build-keys-from-ts.js
Comment (nzakas) [inline tools/build-keys-from-ts.js:25]: Just wondering if there’s a way to avoid having this same code in three places?

Comment (nzakas) [reply tools/build-keys-from-ts.js:25]: I’m just not a fan of needing to update three places if any one of them changes. It’s easy to lose track of that.

Comment (brettz9) [reply tools/build-keys-from-ts.js:25]: I can just create a separate file and import it if you like. It will still be in two places however, as one is the build result, but that's better than three I imagine.

Comment (nzakas) [reply tools/build-keys-from-ts.js:25]: Agreed. My main concern is just people forgetting to update. If one is autogenerated, then it’s not a concern for that one.

Comment (brettz9) [reply tools/build-keys-from-ts.js]: Commit added...

File: tools/backwardCompatibleKeys.js
Comment (nzakas) [inline tools/backwardCompatibleKeys.js:1]: Can we rename this file to backward-compatible-keys.js to match our file naming conventions?

**Review by nzakas — CHANGES_REQUESTED** [review]

File: tools/backwardCompatibleKeys.js
Comment (brettz9) [reply tools/backwardCompatibleKeys.js]: Apologies. Done...

**Review by nzakas — DISMISSED** [review]
LGTM. Thanks! Just want another set of eyes before merging.

**Comment by nzakas** [comment]
Oops, forgot this was a WIP. Whenever you’re ready, feel free to submit for approval.

**Comment by brettz9** [comment]
@nzakas The PR is indeed close to being ready (mostly just need some tests), but really feel I need a decision on https://github.com/eslint/eslint-visitor-keys/pull/36#issuecomment-1049553503 first. 

To summarize that admittedly long comment, we need to decide between the approaches to:
1. Blacklist undesired properties through our build routine only
2. Blacklist undesired properties by adding to the key blacklist of `getKeys`
3. Suppress specific interfaces/types which we know hold the properties we don't want

I supported a mix of algorithms in the event we needed them, but I think as per that comment, we probably only want one (I guess the second one).

Once I know which algorithm to use, I can adapt the code and add tests.

**Comment by nzakas** [comment]
I think the second one makes sense 👍

**Comment by brettz9** [comment]
Ok, it should now be ready.

**Comment by mdjermanovic** [comment]
How should we tag this change? It shouldn't be just a `chore` because it includes some changes in the functionalities of this package:

* It adds `JSXSpreadChild` (that alone would be `feat` ?)
* It changes the public  `getKeys()` (breaking `feat!` ?)

**Comment by brettz9** [comment]
Might the `getKeys` breaking change, be more of a fix? I would think that this should have been already excluding those properties (`loc`, `comments`, `innerComments`, and `operator`).

**Comment by mdjermanovic** [comment]
> Might the `getKeys` breaking change, be more of a fix? I would think that this should have been already excluding those properties (`loc`, `comments`, `innerComments`, and `operator`).

This is a separate package, so it might be correct to do a major version bump when a change seems potentially breaking for packages that depend on it even if the change isn't breaking for ESLint.

As for ESLint, I think everything related to comments can be classified as a fix. They're objects with a `"type"` property, so they would be mistakenly [traversed as nodes](https://github.com/eslint/eslint/blob/8933fe7afcc7cdd99cc0efccc08e8fe3a5e2996f/lib/shared/traverser.js#L33-L35) if properties that point to them were not excluded.

`loc` makes sense as a performance improvement since it's a known ESTree property that exists on all nodes and doesn't point to other nodes. Maybe it would make sense, for the same reason, to also add `type` (I'm not sure why it wasn't added before, and even the [README example](https://github.com/eslint/eslint-visitor-keys#evkgetkeysnode) shows that `getKeys()` returns `"type"`), and `range` although it isn't in the ESTree spec.

`operator` seems like an outlier - other excluded properties are either known properties that have a specific meaning (`parent`, `loc`) or properties related to comments (`comments`, `leadingComments`, `innerComments`, `trailingComments`) . I'm not sure if a generic fallback function should assume that `operator` child is not a node.

**Comment by brettz9** [comment]
> They're objects with a `"type"` property, so they would be mistakenly [traversed as nodes](https://github.com/eslint/eslint/blob/8933fe7afcc7cdd99cc0efccc08e8fe3a5e2996f/lib/shared/traverser.js#L33-L35) if properties that point to them were not excluded.

Ah, there's the missing ingredient, thanks! I had been so focused on the AST, I neglected to dig into the traversal routine to discover the exact specification of the items to be iterated.

> > Might the `getKeys` breaking change, be more of a fix? I would think that this should have been already excluding those properties (`loc`, `comments`, `innerComments`, and `operator`).
>
> This is a separate package, so it might be correct to do a major version bump when a change seems potentially breaking for packages that depend on it even if the change isn't breaking for ESLint.

Ok, in that case, and based on the information above, whether or not we add the excluded keys later to `getKeys` in some breaking release for the purposes of optimization and clarity, I think I will need to tweak the algorithm to exclude properties which do not point to objects with a `type` or those whose type is "Line" or "Comment". That will also take care of `operator` which resolves to a union of strings instead of any object. My existing routine should help, but will take some time to add this detection.

> `loc` makes sense as a performance improvement since it's a known ESTree property that exists on all nodes and doesn't point to other nodes.

Ok. I guess we need a new issue then, especially since I don't actually need to add to `getKeys` in this PR.

> Maybe it would make sense, for the same reason, to also add `type` (I'm not sure why it wasn't added before, and even the [README example](https://github.com/eslint/eslint-visitor-keys#evkgetkeysnode) shows that `getKeys()` returns `"type"`), and `range` although it isn't in the ESTree spec.

Yeah, I was wondering that re: `type` too. And yes, `range`.
 
> `operator` seems like an outlier - other excluded properties are either known properties that have a specific meaning (`parent`, `loc`) or properties related to comments (`comments`, `leadingComments`, `innerComments`, `trailingComments`) . I'm not sure if a generic fallback function should assume that `operator` child is not a node.

Ok, sure. As per the above, I should be able to get the algorithm to detect the non-traversability of `operator` (usually properties with independent types were traversable but not in this case).

**Comment by brettz9** [comment]
Ok, the detection should now be ready without any breaking changes. But `JSXSpreadChild` is added, so indeed apparently a `feat` as discussed.

**Comment by nzakas** [comment]
@mdjermanovic can you take a look?

File: lib/index.js
Comment (mdjermanovic) [inline lib/index.js:25]: Can we remove these lines? We usually avoid having commented-out code in the codebase unless it is associated with a TODO action.

File: tools/get-keys-from-ts.js
Comment (mdjermanovic) [inline tools/get-keys-from-ts.js:213]: Does this function need to be `async`?

File: lib/index.js
Comment (brettz9) [reply lib/index.js:25]: Sure. I had thought these were going to be removed in a future version, so should I add a comment to do so? File an issue? Is there any particular approach for planning ahead with to-dos to occur before breaking changes?

`eslint-plugin-unicorn` has a truly excellent `expiring-todo-comments` rule which can be tied to version or dependency changes which will cause linting to fail unless applied. I might recommend unicorn for `eslint-config-eslint` anyways.

File: tools/get-keys-from-ts.js
Comment (brettz9) [reply tools/get-keys-from-ts.js]: Actually, no. Removed and changed the return type. `require-await` would be a good rule to be adding to `eslint-config-eslint` too I think.

File: lib/index.js
Comment (mdjermanovic) [reply lib/index.js:25]: I think it would be best to discuss this in an issue first.

File: tools/get-keys-from-ts.js
Comment (brettz9) [reply tools/get-keys-from-ts.js:539]: It technically is, and it comes at the cost of an extra microtask, but as per the [no-return-await](https://eslint.org/docs/rules/no-return-await) rule:

> You can avoid the extra microtask by not awaiting the return value, with the trade off of the function no longer being a part of the stack trace if an error is thrown asynchronously from the Promise being returned. This can make debugging more difficult.

File: lib/index.js
Comment (brettz9) [reply lib/index.js]: Ok, thanks for the feedback. Added #38 and removed the comments.

And FWIW, re: Unicorn, I filed https://github.com/eslint/eslint/pull/15731 . Even when not heavily using to-dos in code, having those conditional to-dos can really help ensure that version-dependent to-dos don't get lost.

File: tools/get-keys-from-ts.js
Comment (mdjermanovic) [reply tools/get-keys-from-ts.js:539]: Yes, but `getKeysFromTs` is no longer `async` so `getKeysFromTs(code, options)` is not a Promise?

Comment (brettz9) [reply tools/get-keys-from-ts.js]: 🤦  Yeah :-)

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Review by nzakas — APPROVED** [review]
LGTM. Thanks for all the work on this.

**Pull request merged**
